relay-cogs/src/cogs.rs

```diff
@@ -0,0 +1,336 @@
+    pub fn cancel(&mut self) {
+        // No features -> nothing gets attributed.
+        self.update(AppFeatures::none());
+    }
+
+    /// Updates the app features to whicht the active measurement is attributed to.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// # use relay_cogs::{AppFeature, Cogs, ResourceId};
+    /// # struct Item;
+    /// # fn do_something(_: &Item) -> bool { true };
+    ///
+    /// fn process(cogs: &Cogs, item: &Item) {
+    ///     let mut token = cogs.timed(ResourceId::Relay, AppFeature::Unattributed);
+    ///
+    ///     // App feature is only known after some computations.
+    ///     if do_something(item) {
+    ///         token.update(AppFeature::Spans);
+    ///     } else {
+    ///         token.update(AppFeature::Transactions);
+    ///     }
+    /// }
+    /// ```
+    pub fn update<T: Into<AppFeatures>>(&mut self, features: T) {
+        self.features = features.into();
+    }
+}
+
+impl Drop for CogsToken {
+    fn drop(&mut self) {
+        let elapsed = self.start.elapsed();
+
+        for (feature, ratio) in self.features.weights() {
+            let time = elapsed.div_f32(ratio);
+            self.recorder.record(CogsMeasurement {
+                resource: self.resource,
+                feature,
+                value: Value::Time(time),
+            });
+        }
+    }
+}
+
+impl fmt::Debug for CogsToken {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("CogsToken")
+            .field("resource", &self.resource)
+            .field("features", &self.features)
+            .finish()
+    }
+}
+
+/// A collection of weighted [app features](AppFeature).
+///
+/// Used to attribute a single COGS measurement to multiple features.
+#[derive(Clone, Copy)]
+pub struct AppFeatures(EnumMap<16, AppFeature, NonZeroUsize>);
+
+impl AppFeatures {
+    /// Attributes all measurements to a single [`AppFeature`].
+    pub fn new(feature: AppFeature) -> Self {
+        Self::builder().weight(feature, 1).build()
+    }
+
+    /// Attributes the measurement to nothing.
+    pub fn none() -> Self {
+        Self::builder().build()
+    }
+
+    /// Returns an [`AppFeatures`] builder.
+    pub fn builder() -> AppFeaturesBuilder {
+        AppFeaturesBuilder(Self(Default::default()))
+    }
+
+    /// Merges two instances of [`AppFeatures`] and sums the contained weights.
+    pub fn merge(mut self, other: Self) -> Self {
+        for (feature, weight) in other.0.into_iter() {
+            if let Some(w) = self.0.get_mut(feature) {
+                *w = w.saturating_add(weight.get());
+            } else {
+                self.0.insert(feature, weight);
+            }
+        }
+
+        self
+    }
+
+    /// Returns an iterator yielding an app feature and it's associated weight
+    /// normalized to the total stored weights in the range between `0.0` and `1.0`.
+    ///
+    /// Used to divide a measurement by the stored weights.
```

Comment:
nit: [docs](https://www.notion.so/sentry/Relay-Code-and-Architecture-4fac481cf27741aba54d39594f1cae58?pvs=4#c66b1ab908674eb1a8ca25f3e7f6e423) should start with a single first line.